State serialization and radio communication handling functions

AlexTheDragon8 opened the pull request:
- Wrote functions to serialize and deserialize the rocket's state
- Wrote a function to create a radio command message based on a rocket state
- Wrote a function to compute a checksum based on a rocket state message
- Wrote a function to receive and decode radio command messages character-by-character, including checksum validation

akmorrison committed:
Create unit tests for Alex Kitaev

To run the unit tests, run make in tests/ directory.

AlexTheDragon8 committed:
Added rocket state serialization/deserialization functions

AlexTheDragon8 committed:
Changed serialization/deserialization functions to use a Base64-like encoding

AlexTheDragon8 committed:
Allocated 4 bits for the number of connected boards

AlexTheDragon8 committed:
Fixed bit numbers in comments

akmorrison committed:
Add unit tests for creating a state command

These tests fail. Please make them not fail.

AlexTheDragon8 committed:
Added radio communication handling code

akmorrison commented on radio_handler.c:
```diff
@@ -0,0 +1,63 @@
+        }
+        // Otherwise, simply discard the character.
+    }
+    //TODO
```
Is there anything left to do, or is this just an old comment?

AlexTheDragon8 replied:
Fixed in ae3c242.

akmorrison commented on radio_handler.c:
```diff
@@ -0,0 +1,63 @@
+    static char message[STATE_COMMAND_LEN] = {0};
+    static uint8_t chars_received = 0;
+
+    if(c == '{') {
```
`{` feels like a magic character, can you define it as STATE_COMMAND_HEADER or something?

AlexTheDragon8 replied:
Fixed in ae3c242.

akmorrison commented on radio_handler.h:
```diff
@@ -0,0 +1,36 @@
+ *
+ * If we haven't heard from the operator in TIME_NO_CONTACT_BEFORE_SAFE_STATE,
+ * this function will return VALVE_OPEN, because that's the safest state for the
+ * vent valve to be in
```
Please include a check for how long it's been since we received a full state command over the radio, and if it's been longer than TIME_NO_CONTACT_BEFORE_SAFE_STATE, this should return VALVE_OPEN.

There's a `millis()` function on master that returns milliseconds since bootup. If you merge master into this branch you can use that function

AlexTheDragon8 replied:
Fixed in 8c79fd3.

akmorrison commented on serialize.c:
```diff
@@ -0,0 +1,111 @@
+    if(binary <= 51) return (binary - 26) + 'a';
+    if(binary <= 61) return (binary - 52) + '0';
+    if(binary == 62) return '&';
+    return '/';
```
Can you add a check for if binary is greater than 63, and is thus illegal? I don't know how to handle that error currently, but we shouldn't just silently return `'/'` as we do here

AlexTheDragon8 replied:
What should the code do if binary is greater than 63?

akmorrison replied:
In a different branch we have a report_error function that will be called whenever a programming error like that shows up. Once both branches are merged to master we’ll call that, for now just leave a comment reminding us to report the problem. And return 0 or something

AlexTheDragon8 replied:
Fixed in 049a07c.

akmorrison commented on serialize.c:
```diff
@@ -0,0 +1,111 @@
+    if('a' <= base64 && base64 <= 'z') return base64 - 'a' + 26;
+    if('0' <= base64 && base64 <= '9') return base64 - '0' + 52;
+    if(base64 == '&') return 62;
+    return 63;
```
same here, what should we do if this function receives an illegal character?

AlexTheDragon8 replied:
Fixed in 049a07c.

akmorrison commented on serialize.c:
```diff
@@ -0,0 +1,111 @@
+
+bool serialize_state(const system_state *state, char *str) {
+    if(!state) return false;
+    if(!str) return false;
```
this is a personal best practice thing, but you should test `state != NULL`, not `!state`. C doesn't guarantee that null pointers evaluate to 0.

AlexTheDragon8 replied:
Fixed in ae3c242.

akmorrison commented on serialize.c:
```diff
@@ -0,0 +1,111 @@
+
+bool deserialize_state(system_state *state, const char *str) {
+    if(!state) return false;
+    if(!str) return false;
```
same NULL check here. You may need to `#include <stddef.h>` for NULL to be defined

AlexTheDragon8 replied:
Fixed in ae3c242.

akmorrison commented on serialize.c:
```diff
@@ -0,0 +1,111 @@
+
+bool create_state_command(char *cmd, const system_state *state) {
+    if(cmd == 0) return false;
+    if(state == 0) return false;
```
NULL, not 0

AlexTheDragon8 replied:
Fixed in ae3c242.

akmorrison commented on serialize.c:
```diff
@@ -0,0 +1,111 @@
+    memcpy(cmd + 1, serialized, SERIALIZED_OUTPUT_LEN - 1);
+
+    // Message start indicator
+    cmd[0] = '{';
```
same thing, define `'{'` to be something

AlexTheDragon8 replied:
Fixed in ae3c242.

akmorrison commented on serialize.c:
```diff
@@ -0,0 +1,111 @@
+    uint8_t idx = 0;
+    while(cmd[idx] != 0) {
+        char curr = cmd[idx];
+        uint8_t odd_sum = 0b10101010 & curr, even_sum = 0b01010101 & curr;
```
please put these on separate lines

AlexTheDragon8 replied:
Fixed in ae3c242.

akmorrison reviewed (changes requested):
First impressions. Mostly looks good, and it does pass all the tests. Just fix the minor things

AlexTheDragon8 committed:
- Changed all functions to compare input pointers to NULL rather than 0
 - Added STATE_COMMAND_HEADER macro to represent the { character
 - Removed outdated comments
 - Fixed code style

AlexTheDragon8 committed:
Merge branch 'master' into serialize_state

AlexTheDragon8 committed:
Default to opening the vent valve if no message has been received in a while

AlexTheDragon8 committed:
Added error checks to Base64 encoding/decoding functions

AlexTheDragon8 committed:
Fixed checksum algorithm

AlexTheDragon8 committed:
Removed comments mistakenly left in

akmorrison committed:
Merge branch 'event_loop' into serialize_state

akmorrison committed:
Add error serialize tests

akmorrison commented on serialize.c:
```diff
@@ -0,0 +1,131 @@
+        for(uint8_t i = 0; i < 4; ++i) {
+            odd_sum += 0b00000010 & curr;
+            even_sum += 0b00000001 & curr;
+            curr >> 2;
```
This expression generates no code

AlexTheDragon8 replied:
Fixed in ca0b852.

AlexTheDragon8 committed:
Added functions to serialize and deserialize error messages
Added functions to maintain a ring buffer or error messages to be sent

AlexTheDragon8 committed:
Fixed incorrect line of code in checksum function

akmorrison commented on error.c:
```diff
@@ -1,14 +1,47 @@
 {
-    //do nothing
+    if(err_msg_ring_buf == NULL) {
+        err_msg_ring_buf = malloc(
```
In real time or embedded systems, use of dynamic allocation is highly discouraged. What we do with all of our other queues is to set up a static array of bytes (usually 100-200 of them), and then use that. If the array gets full, we just drop the message. See https://github.com/waterloo-rocketry/canlib/blob/master/util/can_rcv_buffer.c for an example.

I recommend that you use the safe_ring_buffer module from canlib (you may have to go into your canlib directory and pull from master to get it), but if you want to use your own data structure, you shouldn't use malloc

akmorrison commented on error.h:
```diff
@@ -3,10 +3,14 @@
+// How many error messages can be buffered at once
+#define ERROR_MESSAGE_RING_BUFFER_SIZE 16
 
 typedef struct {
```
I don't think any other code modules are going to use this type, it can probably be moved to the c file instead of the h file

akmorrison reviewed (changes requested):
The serialize functions look perfect, but the queueing of error messages shouldn't use `malloc`

AlexTheDragon8 committed:
Changed error message ring buffer to use a statically allocated array

akmorrison commented on error.c:
```diff
@@ -1,9 +1,41 @@
+    char *serialized_error = err_msg_ring_buf[err_msg_buf_write];
+    serialize_error(&err, serialized_error);
+    ++err_msg_buf_size;
+    ++err_msg_buf_write;
```
if err_msg_bug_write exceeds ERROR_MESSAGE_RING_BUFFER_SIZE, what happens here?

AlexTheDragon8 replied:
Fixed in 1cf2ee2.

akmorrison commented on error.c:
```diff
@@ -1,9 +1,41 @@
+    char *serialized_error = err_msg_ring_buf[err_msg_buf_read];
+    memcpy(output, serialized_error, ERROR_COMMAND_LENGTH);
+
+    ++err_msg_buf_read;
```
This isn't really a ring buffer, on account of it not wrapping around when it gets to the end

AlexTheDragon8 replied:
Fixed in 1cf2ee2.

akmorrison reviewed (changes requested):
I think your ring buffer is missing the check for writing past the end of the array. You could get away with just using safe_ring_buffer...

akmorrison committed:
Add randomized test to error serializer

Make sure report_error can handle 1000 errors. Not all at the same time,
but eventually.

AlexTheDragon8 committed:
Fixed ring buffer positions not wrapping around the buffer

AlexTheDragon8 committed:
Included string.h for memcpy()

AlexTheDragon8 committed:
Included serialize.h in error.c to avoid implicit declaration
